codegen-server/python/src/main/kotlin/software/amazon/smithy/rust/codegen/server/python/smithy/PythonServerSymbolProvider.kt

```diff
@@ -3,14 +3,91 @@
 import software.amazon.smithy.rust.codegen.smithy.RustSymbolProvider
-import software.amazon.smithy.rust.codegen.smithy.WrappingSymbolProvider
+import software.amazon.smithy.rust.codegen.smithy.SHAPE_KEY
+import software.amazon.smithy.rust.codegen.smithy.SymbolVisitorConfig
 import software.amazon.smithy.rust.codegen.smithy.rustType
+import software.amazon.smithy.rust.codegen.util.toPascalCase
+import software.amazon.smithy.rust.codegen.util.toSnakeCase
+
+// Wrapping symbol visitor provider allowing to implement symbol providers that can recursively replace symbols in nested shapes.
+open class PythonWrappingVisitingSymbolProvider(private val base: RustSymbolProvider, private val model: Model) : ShapeVisitor.Default<Symbol>(), RustSymbolProvider {
+    override fun getDefault(shape: Shape): Symbol {
+        return base.toSymbol(shape)
+    }
+
+    override fun config(): SymbolVisitorConfig {
+        return base.config()
+    }
+
+    override fun toEnumVariantName(definition: EnumDefinition): MaybeRenamed? {
+        return base.toEnumVariantName(definition)
+    }
+
+    override fun toMemberName(shape: MemberShape): String = when (val container = model.expectShape(shape.container)) {
+        is StructureShape -> shape.memberName.toSnakeCase()
+        is UnionShape -> shape.memberName.toPascalCase()
+        else -> error("unexpected container shape: $container")
+    }
+
+    override fun listShape(shape: ListShape): Symbol {
+        val inner = toSymbol(shape.member)
+        return symbolBuilder(shape, RustType.Vec(inner.rustType())).addReference(inner).build()
+    }
+
+    override fun mapShape(shape: MapShape): Symbol {
+        val target = model.expectShape(shape.key.target)
+        require(target.isStringShape) { "unexpected key shape: ${shape.key}: $target [keys must be strings]" }
+        val key = toSymbol(shape.key)
+        val value = toSymbol(shape.value)
+        return symbolBuilder(shape, RustType.HashMap(key.rustType(), value.rustType())).addReference(key)
+            .addReference(value).build()
+    }
+
+    override fun setShape(shape: SetShape): Symbol {
+        val inner = toSymbol(shape.member)
+        val builder = if (model.expectShape(shape.member.target).isStringShape) {
+            symbolBuilder(shape, RustType.HashSet(inner.rustType()))
+        } else {
+            // only strings get put into actual sets because floats are unhashable
+            symbolBuilder(shape, RustType.Vec(inner.rustType()))
+        }
+        return builder.addReference(inner).build()
+    }
+
+    override fun memberShape(shape: MemberShape): Symbol {
+        return toSymbol(model.expectShape(shape.target))
+    }
+
+    override fun toSymbol(shape: Shape): Symbol {
+        return shape.accept(this)
+    }
+
+    private fun symbolBuilder(shape: Shape?, rustType: RustType): Symbol.Builder {
+        val builder = Symbol.builder().putProperty("shape", SHAPE_KEY)
```

Comment:
Lol. To be honest I did not understand this at all. Thanks for pointing it out. I think you are right, this is not used at all. I'll validate and refactor this.

Reply:
I have added the `.shape` method to have a nicer interface. I don't think this is used anywhere thou, but I am not sure about the impact of just removing it. 

I added an issue (#1514) to track this, since I don't believe it should be done as part of the python PR.

Reply:
I have deprecated it, this PR will also close #1514